File: packages/spectrum-managed/src/wire.js
// src/wire.js — SHP1 zero-copy decoder for weft_hw_profile_t (Pillar 5).
//
// Normative layout: docs/spectrum/SPECTRUM-WIRE-V1.md (byte-frozen).
// Law 1: ProfileView is constructed ONCE over an existing buffer; every
// getter returns a primitive read in place — the steady-state telemetry loop
// allocates NOTHING. snapshotInto() copies primitives into a caller-owned
// flyweight (also allocation-free).
// Law 2: all multi-byte reads are explicit little-endian DataView ops at
// frozen offsets. Law 4: integrity violations surface as integer codes
// (§6 taxonomy) — a torn read can NEVER silently decode garbage.

export const RECORD_SIZE = 192;
export const CRC_OFFSET = 188;
export const MAGIC0 = 0x53; // 'S'
export const MAGIC1 = 0x48; // 'H'
export const MAGIC2 = 0x50; // 'P'
export const MAGIC3 = 0x31; // '1'
export const LAYOUT_VERSION = 1;

// §6 Law 4 error taxonomy (frozen codes; shared across all managed runtimes)
export const E_BAD_MAGIC = 1;
export const E_BAD_VERSION = 2;
export const E_BAD_SIZE = 3;
export const E_CRC_MISMATCH = 4;
export const E_RESERVED_DIRTY = 5;
export const E_PROBE_UNAVAILABLE = 6;
export const E_DEVICE_LOST = 7;
export const E_FFI_TIMEOUT = 8;
export const E_HEAP_PRESSURE = 9;
export const E_LISTENER_LEAK = 10;
export const E_ALIGN_INVALID = 11;
export const E_HUD_CONTEXT_LOST = 12;
export const E_UNMARSHAL_FAILED = 13;
export const E_TIER_EXHAUSTED = 14;
export const E_HUD_RECOVERED = 15;

export const ERROR_NAMES = new Map([
  [E_BAD_MAGIC, 'E_BAD_MAGIC'], [E_BAD_VERSION, 'E_BAD_VERSION'],
  [E_BAD_SIZE, 'E_BAD_SIZE'], [E_CRC_MISMATCH, 'E_CRC_MISMATCH'],
  [E_RESERVED_DIRTY, 'E_RESERVED_DIRTY'], [E_PROBE_UNAVAILABLE, 'E_PROBE_UNAVAILABLE'],
  [E_DEVICE_LOST, 'E_DEVICE_LOST'], [E_FFI_TIMEOUT, 'E_FFI_TIMEOUT'],
  [E_HEAP_PRESSURE, 'E_HEAP_PRESSURE'], [E_LISTENER_LEAK, 'E_LISTENER_LEAK'],
  [E_ALIGN_INVALID, 'E_ALIGN_INVALID'], [E_HUD_CONTEXT_LOST, 'E_HUD_CONTEXT_LOST'],
  [E_UNMARSHAL_FAILED, 'E_UNMARSHAL_FAILED'], [E_TIER_EXHAUSTED, 'E_TIER_EXHAUSTED'],
  [E_HUD_RECOVERED, 'E_HUD_RECOVERED'],
]);

// Enum domains (§2)
export const TIER_UNKNOWN = 0, TIER_FLAGSHIP = 1, TIER_MID = 2, TIER_BUDGET = 3;
export const THERMAL_NOMINAL = 0, THERMAL_LIGHT = 1, THERMAL_MODERATE = 2,
  THERMAL_SEVERE = 3, THERMAL_CRITICAL = 4;
export const VIS_VISIBLE = 0, VIS_HIDDEN = 1, VIS_UNKNOWN = 2;
export const CHARGING_NO = 0, CHARGING_YES = 1, CHARGING_UNKNOWN = 2;
export const BATTERY_UNKNOWN = 0xffff;

// §3 feature bits (u64 as lo/hi u32 pair)
export const FEAT = {
  WASM_SIMD128: 0, SHARED_ARRAY_BUFFER: 1, WEBGPU: 2, WEBGL2: 3,
  AVX512: 4, AVX2: 5, SSE42: 6, NEON: 7, SVE2: 8, RVV: 9,
  METAL_3: 10, CUDA: 11, APPLE_MPS: 12, OPENVINO: 13,
  FASTRPC_DSP: 14, NEUROPILOT: 15, MULTILANE_DMA: 16, BIG_LITTLE: 17,
  THERMAL_SENSOR: 18, DLPACK_EXPORT: 19,
};
const FEAT_BY_NAME = new Map(Object.entries(FEAT).map(([k, v]) => [v, k]));

// CRC-32 (IEEE 802.3, reflected, init/final 0xFFFFFFFF) — identical arithmetic
// to fixtures/generate.mjs, python/weft_spectrum/wire.py and the native audits.
const CRC_TABLE = new Int32Array(256);
for (let n = 0; n < 256; n++) {
  let c = n;
  for (let k = 0; k < 8; k++) c = (c & 1) !== 0 ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1);
  CRC_TABLE[n] = c;
}

export function crc32(bytes, len) {
  let c = -1;
  for (let i = 0; i < len; i++) c = CRC_TABLE[(c ^ bytes[i]) & 0xFF] ^ (c >>> 8);
  return (c ^ -1) >>> 0;
}

// ---------------------------------------------------------------------------
// ProfileView — zero-copy window over ONE 192-byte record.
// ---------------------------------------------------------------------------
const U64_FIELDS = [[40, 'cpuMaxClockKhz'], [48, 'memoryTotalBytes'], [56, 'memoryBudgetBytes'], [72, 'maxFrameRateMilliHz']];

export class ProfileView {
  constructor(buffer, byteOffset = 0) {
    if (buffer instanceof DataView) {
      this._dv = buffer;
    } else if (ArrayBuffer.isView(buffer)) {
      // TypedArray/Buffer input (Buffer is a view over a pooled ArrayBuffer)
      this._dv = new DataView(buffer.buffer, buffer.byteOffset + byteOffset, buffer.byteLength);
    } else {
      this._dv = new DataView(buffer, byteOffset, RECORD_SIZE);
    }
    this._bytes = new Uint8Array(this._dv.buffer, this._dv.byteOffset, this._dv.byteLength);
  }

  // Law 4 gate: returns 0 when the record is intact, else the §6 code.
  // Checks size -> magic -> version -> reserved -> CRC (cheapest first).
  validate() {
    if (this._dv.byteLength < RECORD_SIZE) return E_BAD_SIZE;
    if (this._bytes[0] !== MAGIC0 || this._bytes[1] !== MAGIC1 ||
      this._bytes[2] !== MAGIC2 || this._bytes[3] !== MAGIC3) return E_BAD_MAGIC;
    if (this._dv.getUint16(4, true) !== LAYOUT_VERSION) return E_BAD_VERSION;
    if (this._dv.getUint16(6, true) !== RECORD_SIZE) return E_BAD_SIZE;
    for (let i = 104; i < CRC_OFFSET; i++) if (this._bytes[i] !== 0) return E_RESERVED_DIRTY;
    if (crc32(this._bytes, CRC_OFFSET) !== this._dv.getUint32(CRC_OFFSET, true)) return E_CRC_MISMATCH;
    return 0;
  }

  get layoutVersion() { return this._dv.getUint16(4, true); }
  get recordSize() { return this._dv.getUint16(6, true); }
  get featureFlagsLo() { return this._dv.getUint32(8, true); }
  get featureFlagsHi() { return this._dv.getUint32(12, true); }
  get siliconTier() { return this._dv.getUint32(16, true); }
  get thermalState() { return this._dv.getUint32(20, true); }
  get perfCores() { return this._dv.getUint32(24, true); }
  get effCores() { return this._dv.getUint32(28, true); }
  get gpuFamily() { return this._dv.getUint32(32, true); }
  get cacheLineBytes() { return this._dv.getUint32(36, true); }
  get simdWidthBits() { return this._dv.getUint32(64, true); }
  get frameBudgetUs() { return this._dv.getUint32(68, true); }
  get batteryPermille() { return this._dv.getUint32(80, true); }
  get batteryCharging() { return this._dv.getUint32(84, true); }
  get visibility() { return this._dv.getUint32(88, true); }
  get dmaLaneCount() { return this._dv.getUint32(92, true); }
  get vendorId() { return this._dv.getUint32(96, true); }
  get deviceId() { return this._dv.getUint32(100, true); }
  get crc32() { return this._dv.getUint32(CRC_OFFSET, true); }
  get cpuMaxClockKhz() { return this._dv.getUint32(40, true) + this._dv.getUint32(44, true) * 0x100000000; }
  get memoryTotalBytes() { return this._dv.getUint32(48, true) + this._dv.getUint32(52, true) * 0x100000000; }
  get memoryBudgetBytes() { return this._dv.getUint32(56, true) + this._dv.getUint32(60, true) * 0x100000000; }
  get maxFrameRateMilliHz() { return this._dv.getUint32(72, true) + this._dv.getUint32(76, true) * 0x100000000; }

  // 64-bit reads use lo/hi u32 pairs (no BigInt allocs on the hot path)
  u64(field) {
    for (let i = 0; i < U64_FIELDS.length; i++) {
      if (U64_FIELDS[i][1] === field) {
        const off = U64_FIELDS[i][0];
        return this._dv.getUint32(off, true) + this._dv.getUint32(off + 4, true) * 0x100000000;
      }
    }
    return 0;
  }

  hasFeatureBit(bit) {
    return bit < 32
      ? (this.featureFlagsLo & ((1 >>> 0) << bit)) !== 0
      : (this.featureFlagsHi & ((1 >>> 0) << (bit - 32))) !== 0;
  }
  featureName(bit) { return FEAT_BY_NAME.get(bit) || null; }

  // Copies all fields into a caller-owned flyweight (Law 1: no allocation).
  snapshotInto(dst) {
    dst.layoutVersion = this.layoutVersion;
    dst.recordSize = this.recordSize;
    dst.featureFlagsLo = this.featureFlagsLo;
    dst.featureFlagsHi = this.featureFlagsHi;
    dst.siliconTier = this.siliconTier;
    dst.thermalState = this.thermalState;
    dst.perfCores = this.perfCores;
    dst.effCores = this.effCores;
    dst.gpuFamily = this.gpuFamily;
    dst.cacheLineBytes = this.cacheLineBytes;
    dst.cpuMaxClockKhz = this.cpuMaxClockKhz;
    dst.memoryTotalBytes = this.memoryTotalBytes;
    dst.memoryBudgetBytes = this.memoryBudgetBytes;
    dst.simdWidthBits = this.simdWidthBits;
    dst.frameBudgetUs = this.frameBudgetUs;
    dst.maxFrameRateMilliHz = this.maxFrameRateMilliHz;
    dst.batteryPermille = this.batteryPermille;
    dst.batteryCharging = this.batteryCharging;
    dst.visibility = this.visibility;
    dst.dmaLaneCount = this.dmaLaneCount;
    dst.vendorId = this.vendorId;
    dst.deviceId = this.deviceId;
    return dst;
  }
}

export function makeProfileFlyweight() {
  return {
    layoutVersion: 0, recordSize: 0, featureFlagsLo: 0, featureFlagsHi: 0,
    siliconTier: 0, thermalState: 0, perfCores: 0, effCores: 0, gpuFamily: 0,
    cacheLineBytes: 64, cpuMaxClockKhz: 0, memoryTotalBytes: 0,
    memoryBudgetBytes: 0, simdWidthBits: 0, frameBudgetUs: 0,
    maxFrameRateMilliHz: 0, batteryPermille: BATTERY_UNKNOWN,
    batteryCharging: CHARGING_UNKNOWN, visibility: VIS_UNKNOWN,
    dmaLaneCount: 0, vendorId: 0, deviceId: 0,
  };
}

// Convenience: validate + snapshot in one call. Init path (may allocate the
// result object); steady-state polling should use ProfileView directly.
export function decodeProfile(buffer, byteOffset = 0) {
  const view = new ProfileView(buffer, byteOffset);
  const code = view.validate();
  if (code !== 0) return { ok: false, code, name: ERROR_NAMES.get(code), view };
  return { ok: true, code: 0, name: null, view };
}

File: python/weft_spectrum/weft_spectrum/wire.py
# weft_spectrum/wire.py — SHP1 zero-copy decoder for weft_hw_profile_t (P5).
#
# Normative layout: docs/spectrum/SPECTRUM-WIRE-V1.md (byte-frozen, shared
# with packages/spectrum-managed/src/wire.js — offsets MUST stay identical;
# ci stage 1 + tests/spectrum/managed/parity.mjs enforce this).
#
# Law 1: the steady-state loop decodes into a __slots__ flyweight via ONE
# struct.unpack_from call; transient CPython objects are freelist-reused and
# RETAINED heap growth over 1,000,000 cycles is < 64 KiB (probe: tests).
# Law 2: explicit '<' little-endian struct — no native sizing, ever.
# Law 4: integrity violations surface as integer codes; torn records can
# never silently decode as garbage.

import zlib
from struct import Struct

RECORD_SIZE = 192
CRC_OFFSET = 188
MAGIC = b"SHP1"
LAYOUT_VERSION = 1

# §6 Law 4 error taxonomy (frozen; identical to wire.js)
E_BAD_MAGIC = 1
E_BAD_VERSION = 2
E_BAD_SIZE = 3
E_CRC_MISMATCH = 4
E_RESERVED_DIRTY = 5
E_PROBE_UNAVAILABLE = 6
E_DEVICE_LOST = 7
E_FFI_TIMEOUT = 8
E_HEAP_PRESSURE = 9
E_LISTENER_LEAK = 10
E_ALIGN_INVALID = 11
E_HUD_CONTEXT_LOST = 12
E_UNMARSHAL_FAILED = 13
E_TIER_EXHAUSTED = 14
E_HUD_RECOVERED = 15

ERROR_NAMES = {
    1: "E_BAD_MAGIC", 2: "E_BAD_VERSION", 3: "E_BAD_SIZE", 4: "E_CRC_MISMATCH",
    5: "E_RESERVED_DIRTY", 6: "E_PROBE_UNAVAILABLE", 7: "E_DEVICE_LOST",
    8: "E_FFI_TIMEOUT", 9: "E_HEAP_PRESSURE", 10: "E_LISTENER_LEAK",
    11: "E_ALIGN_INVALID", 12: "E_HUD_CONTEXT_LOST", 13: "E_UNMARSHAL_FAILED",
    14: "E_TIER_EXHAUSTED", 15: "E_HUD_RECOVERED",
}

TIER_UNKNOWN, TIER_FLAGSHIP, TIER_MID, TIER_BUDGET = 0, 1, 2, 3
THERMAL_NOMINAL, THERMAL_LIGHT, THERMAL_MODERATE = 0, 1, 2
THERMAL_SEVERE, THERMAL_CRITICAL = 3, 4
VIS_VISIBLE, VIS_HIDDEN, VIS_UNKNOWN = 0, 1, 2
CHARGING_NO, CHARGING_YES, CHARGING_UNKNOWN = 0, 1, 2
BATTERY_UNKNOWN = 0xFFFF

# §3 feature bits (identical to wire.js FEAT)
FEAT = {
    "WASM_SIMD128": 0, "SHARED_ARRAY_BUFFER": 1, "WEBGPU": 2, "WEBGL2": 3,
    "AVX512": 4, "AVX2": 5, "SSE42": 6, "NEON": 7, "SVE2": 8, "RVV": 9,
    "METAL_3": 10, "CUDA": 11, "APPLE_MPS": 12, "OPENVINO": 13,
    "FASTRPC_DSP": 14, "NEUROPILOT": 15, "MULTILANE_DMA": 16,
    "BIG_LITTLE": 17, "THERMAL_SENSOR": 18, "DLPACK_EXPORT": 19,
}
FEAT_BY_BIT = {v: k for k, v in FEAT.items()}

# One-call record decode: '<4s 2H 8I 3Q 2I Q 6I 84x I' = 4+4+32+24+8+8+24+84+4
# = 192 bytes. '<' pins little-endian + no padding (Law 2).
_RECORD = Struct("<4s2H8I3Q2IQ6I84xI")
_RECORD_UNPACK = _RECORD.unpack_from

_RESERVED_ZERO = bytes(84)
_crc32 = zlib.crc32


class ProfileFlyweight:
    """Reusable decode target (Law 1). One instance, mutated in place forever."""
    __slots__ = (
        "layoutVersion", "recordSize", "featureFlagsLo", "featureFlagsHi",
        "siliconTier", "thermalState", "perfCores", "effCores", "gpuFamily",
        "cacheLineBytes", "cpuMaxClockKhz", "memoryTotalBytes",
        "memoryBudgetBytes", "simdWidthBits", "frameBudgetUs",
        "maxFrameRateMilliHz", "batteryPermille", "batteryCharging",
        "visibility", "dmaLaneCount", "vendorId", "deviceId",
    )

    def __init__(self):
        self.layoutVersion = 0
        self.recordSize = 0
        self.featureFlagsLo = 0
        self.featureFlagsHi = 0
        self.siliconTier = TIER_UNKNOWN
        self.thermalState = THERMAL_NOMINAL
        self.perfCores = 0
        self.effCores = 0
        self.gpuFamily = 0
        self.cacheLineBytes = 64
        self.cpuMaxClockKhz = 0
        self.memoryTotalBytes = 0
        self.memoryBudgetBytes = 0
        self.simdWidthBits = 0
        self.frameBudgetUs = 0
        self.maxFrameRateMilliHz = 0
        self.batteryPermille = BATTERY_UNKNOWN
        self.batteryCharging = CHARGING_UNKNOWN
        self.visibility = VIS_UNKNOWN
        self.dmaLaneCount = 0
        self.vendorId = 0
        self.deviceId = 0

    def features(self):
        """Init-time convenience (allocates; NOT for the steady-state loop)."""
        flags = (self.featureFlagsHi << 32) | self.featureFlagsLo
        return [FEAT_BY_BIT[b] for b in sorted(FEAT_BY_BIT) if (flags >> b) & 1]


class ProfileView:
    """Zero-copy window over ONE 192-byte SHP1 record (bytes/bytearray)."""

    __slots__ = ("_buf",)

    def __init__(self, buf):
        self._buf = buf  # zero-copy: holds the caller's buffer, no snapshot

    def validate(self):
        """0 when intact, else a §6 code. Cheapest checks first."""
        buf = self._buf
        if len(buf) < RECORD_SIZE:
            return E_BAD_SIZE
        if bytes(buf[0:4]) != MAGIC:
            return E_BAD_MAGIC
        if (buf[4] | (buf[5] << 8)) != LAYOUT_VERSION:
            return E_BAD_VERSION
        if (buf[6] | (buf[7] << 8)) != RECORD_SIZE:
            return E_BAD_SIZE
        if buf[104:188] != _RESERVED_ZERO:
            return E_RESERVED_DIRTY
        if _crc32(memoryview(buf)[0:CRC_OFFSET]) != (
            buf[188] | (buf[189] << 8) | (buf[190] << 16) | (buf[191] << 24)
        ):
            return E_CRC_MISMATCH
        return 0

    def snapshot_into(self, dst):
        """Decode all fields into the flyweight (returns dst; no new objects)."""
        (magic, version, size, lo, hi, tier, thermal, perf, eff, gpu, cache,
         clock, mem_total, mem_budget, simd, fbudget, max_rate, batt, charg,
         vis, dma, vendor, device, crc) = _RECORD_UNPACK(self._buf, 0)
        dst.layoutVersion = version
        dst.recordSize = size
        dst.featureFlagsLo = lo
        dst.featureFlagsHi = hi
        dst.siliconTier = tier
        dst.thermalState = thermal
        dst.perfCores = perf
        dst.effCores = eff
        dst.gpuFamily = gpu
        dst.cacheLineBytes = cache
        dst.cpuMaxClockKhz = clock
        dst.memoryTotalBytes = mem_total
        dst.memoryBudgetBytes = mem_budget
        dst.simdWidthBits = simd
        dst.frameBudgetUs = fbudget
        dst.maxFrameRateMilliHz = max_rate
        dst.batteryPermille = batt
        dst.batteryCharging = charg
        dst.visibility = vis
        dst.dmaLaneCount = dma
        dst.vendorId = vendor
        dst.deviceId = device
        return dst

    # primitive byte-offset getters (mirrors wire.js offsets exactly)
    @property
    def featureFlagsLo(self):
        return self._buf[8] | (self._buf[9] << 8) | (self._buf[10] << 16) | (self._buf[11] << 24)

    @property
    def featureFlagsHi(self):
        return self._buf[12] | (self._buf[13] << 8) | (self._buf[14] << 16) | (self._buf[15] << 24)

    @property
    def siliconTier(self):
        return self._buf[16] | (self._buf[17] << 8) | (self._buf[18] << 16) | (self._buf[19] << 24)

    @property
    def thermalState(self):
        return self._buf[20] | (self._buf[21] << 8) | (self._buf[22] << 16) | (self._buf[23] << 24)

    @property
    def memoryBudgetBytes(self):
        b = self._buf
        return b[56] | (b[57] << 8) | (b[58] << 16) | (b[59] << 24) | \
            (b[60] << 32) | (b[61] << 40) | (b[62] << 48) | (b[63] << 56)

    @property
    def maxFrameRateMilliHz(self):
        b = self._buf
        return b[72] | (b[73] << 8) | (b[74] << 16) | (b[75] << 24) | \
            (b[76] << 32) | (b[77] << 40) | (b[78] << 48) | (b[79] << 56)

    @property
    def crc32(self):
        b = self._buf
        return b[188] | (b[189] << 8) | (b[190] << 16) | (b[191] << 24)

    def has_feature_bit(self, bit):
        if bit < 32:
            return (self.featureFlagsLo & (1 << bit)) != 0
        return (self.featureFlagsHi & (1 << (bit - 32))) != 0


def crc32_ref(data):
    """Reference CRC (identical arithmetic to wire.js/generate.mjs)."""
    return _crc32(memoryview(data)) & 0xFFFFFFFF


def decode_profile(buf, dst=None):
    """Init-path convenience: validate + snapshot. Returns (code, fw, view)."""
    view = ProfileView(buf)
    code = view.validate()
    if dst is None:
        dst = ProfileFlyweight()
    if code == 0:
        view.snapshot_into(dst)
    return code, dst, view
